torch/_refs/linalg/__init__.py

```diff
@@ -107,29 +108,36 @@ def vector_norm(
     *,
     dtype: Optional[torch.dtype] = None,
 ) -> Tensor:
-    from torch.fx.experimental.symbolic_shapes import guard_size_oblivious
+    from torch.fx.experimental.symbolic_shapes import guard_or_true
 
     # Checks
     check_fp_or_complex(x.dtype, "linalg.vector_norm")
 
     if isinstance(dim, Dim):
         dim = [dim]  # type: ignore[assignment]
 
-    if guard_size_oblivious(x.numel() == 0) and (ord < 0.0 or ord == float("inf")):
-        torch._check(
-            dim is not None and len(dim) != 0,
-            lambda: f"linalg.vector_norm cannot compute the {ord} norm on an empty tensor "
-            "because the operation does not have an identity",
-        )
-        shape = x.shape
-        assert dim is not None  # mypy does not seem to be able to see through check?
-        for d in dim:
+    if ord < 0.0 or ord == float("inf"):
+        # No need to do the checks bellow if we can identify at compile time that
+        # x.numel() != 0.
+        # x.numel() != 0 is repeated in the checks on purpose to handle the situation
+        # where x.numel() == 0 is data dependent. In such case it will be checked at runtime.
+        if guard_or_true(x.numel() == 0):
             torch._check(
-                shape[d] != 0,
-                lambda: f"linalg.vector_norm cannot compute the {ord} norm on the "
-                f"dimension {d} because this dimension is empty and the "
-                "operation does not have an identity",
+                sym_or(x.numel() != 0, dim is not None and len(dim) != 0),
+                lambda: f"linalg.vector_norm cannot compute the {ord} norm on an empty tensor "
+                "because the operation does not have an identity",
             )
+            shape = x.shape
+            assert (
+                dim is not None
+            )  # mypy does not seem to be able to see through check?
```

Comment:
It's true that mypy doesn't understand torch._check() - but also your check isn't guaranteeing that dim is not None, is it? (If `x.numel() != 0` then `dim` could be `None`)

Reply:
this is exiting code i did not add this

Reply:
but anyway i am changing this will add an if